truffle/com.oracle.truffle.api.instrumentation/src/com/oracle/truffle/api/instrumentation/TruffleInstrument.java

```diff
@@ -0,0 +1,236 @@
+ * </p>
+ *
+ * <p>
+ * An implementation must use the {@link Registration} annotation on the class to provide the
+ * necessary meta-data for the class and to enable automatic discovery of the implementation.
+ * </p>
+ *
+ * <b> Example for a simple expression coverage instrumentation: </b>
+ *
+ * <pre>
+ * &#064;Registration(name = Coverage.NAME, version = Coverage.VERSION, instrumentType = Coverage.TYPE)
+ * public final class Coverage extends Instrumentation {
+ *
+ *     public static final String NAME = &quot;sample-coverage&quot;;
+ *     public static final String TYPE = &quot;coverage&quot;;
+ *     public static final String VERSION = &quot;coverage&quot;;
+ *
+ *     private final Set&lt;SourceSection&gt; coverage = new HashSet&lt;&gt;();
+ *
+ *     &#064;Override
+ *     protected void onCreate(Env env, Instrumenter instrumenter) {
+ *         instrumenter.attachFactory(SourceSectionFilter.newBuilder() //
+ *         .tagIs(&quot;EXPRESSION&quot;).build(), new EventNodeFactory() {
+ *             public EventNode create(final EventContext context) {
+ *                 return new EventNode() {
+ *                     &#064;CompilationFinal private boolean visited;
+ *
+ *                     &#064;Override
+ *                     public void onReturnValue(VirtualFrame vFrame, Object result) {
+ *                         if (!visited) {
+ *                             CompilerDirectives.transferToInterpreterAndInvalidate();
+ *                             visited = true;
+ *                             coverage.add(context.getInstrumentedSourceSection());
+ *                         }
+ *                     }
+ *                 };
+ *             }
+ *         });
+ *     }
+ *
+ *     &#064;Override
+ *     protected void onDispose(Env env) {
+ *         // print result
+ *     }
+ *
+ * }
+ * </pre>
+ */
+public abstract class TruffleInstrument {
+
+    /**
+     * Method invoked if the instrumentation is allocated and used by the runtime system. Invoked
+     * exactly once per {@link TruffleInstrument} instance.
+     *
+     * @param env environment information for the instrumentation
+     * @param instrumenter to attach event bindings to the runtime system
+     */
+    protected abstract void onCreate(Env env, Instrumenter instrumenter);
+
+    /**
+     * Method invoked if the instrument is disabled or the underlying engine is disposed. Invoked
+     * exactly once per {@link TruffleInstrument} instance. If the instrument is re-enabled a new
+     * {@link TruffleInstrument} implementation instance is created.
+     *
+     * @param env environment information for the instrumentation
+     */
+    protected void onDispose(Env env) {
+        // default implementation does nothing
+    }
+
+    /**
+     * Represents the environment of a {@link TruffleInstrument}. Provides ways and means to access
+     * input, output and error streams. Also allows to parse arbitrary code from other Truffle
+     * languages.
```

Comment:
First sentence adds little useful information. What kind of environment is it, and what role does it play?  The remaining two sentences just repeat what can be easily observed.  Is there nothing else important about this environment worth discussing?